[xuwen-ho] iP

xuwen-ho opened the pull request:
**HAL 9000 Task Management System**

🌌 *Your mind is for having ideas, not holding them.* – David Allen ([source](https://example.com/source))

> "I'm sorry, Dave. I'm afraid I can't do that." - HAL 9000

1. **Introduction to HAL 9000:**
   - HAL 9000 is an advanced task management system designed to free your mind from the burden of remembering tasks. It's inspired by the iconic AI from the movie *2001: A Space Odyssey*.

2. **Features of HAL 9000:**
   - **Text-Based:** Interact with HAL 9000 through intuitive text commands, just like conversing with an intelligent AI.
   - **Easy to Learn:** HAL's interface is designed to be user-friendly, making task management effortless.
   - **FAST SUPER FAST:** HAL 9000 operates at lightning speed, ensuring swift task handling and responses.

3. **Getting Started with HAL 9000:**
   - Download HAL 9000.
   - Double-click the executable file to launch HAL 9000.
   - Start adding your tasks and let HAL manage them for you. 😉

4. **Main Method for HAL 9000:**
   ```java
   public static void main(String[] args) {
        new Main("data/hal.txt").run();
    }
   ```

- [x] **Managing tasks:** HAL 9000 excels at organizing your tasks efficiently.
- [ ] **Managing deadlines (coming soon):** Stay tuned for upcoming features to manage deadlines seamlessly.
- [ ] **Reminders (coming soon):** HAL 9000 will soon provide reminders to ensure you never miss important tasks.

*If you're a Java programmer, you can also use HAL 9000 to practice Java programming and interact with it programmatically.*

Upgrade to HAL 9000 today and experience a revolutionary way to manage your tasks in the vast expanse of the digital universe! 🚀

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

xuwen-ho committed:
Made chatbot recite its name, greet the user, before saying goodbye

xuwen-ho committed:
Implemented Echo functionality
Implemented Exit functionality upon using 'bye' command

xuwen-ho committed:
Added Storage class
Added feature to add tasks as well as to display tasks

xuwen-ho committed:
Added feature to mark tasks as done or undone

Currently, tasks are only listed with descriptions

This means that we cannot tell if they are done or not, which defeats the purpose of a todoList

This commit will implement such a feature

xuwen-ho committed:
Implmented ToDos, Events, Deadlines to replace generic Tasks

Currently, every task is seen as the same, and one is unable to specify deadlines

This lack of customisability means that it is difficult to see when tasks start and end, or to differentiate tasks

Adding these features makes it clearer which category a task belongs to

xuwen-ho committed:
Edited the Expected Output file from the program

Currently, expected output is the standard boilerplate

This does not reflect the new features added to the chatbot

Through changing the Expected.txt file, allow us to have automated testing

xuwen-ho committed:
Subject: Added Error-Handling capabilities

Current State: No error message displayed when an invalid keyword was provided, or a task added without desciprtion

What was Wrong: This meant that any random keyword could be entered, or invalid tasks could be added

How Commit changes Current State:Error checking to ensure that the above does not take place

xuwen-ho committed:
Subject: Added feature to delete tasks

Current State: Only can add tasks, but unable to delete

What was Wrong: No feature present to remove tasks if they were no longer neccessary

How Commit changes Current State: Allows deletion of redundant tasks

xuwen-ho committed:
Subject: Add feature to store previous todos in system memory

Current State: Todolist starts from empty state each time program is run

What was Wrong: This means we cannot keep track of all our todos, having to add previously added todos again

How Commit changes Current State: Lets store the todos so we do not have to add them again

xuwen-ho committed:
Subject: Imported LocalDate libary to try to implement Date parsing ability

Current State: See date entered as purely a string

What was Wrong: Limits what we can do with entered date, cannot compare etc

How Commit changes Current State: Lets parse the date so we can do other actions like list all task before a certain date

xuwen-ho committed:
Merge branch 'branch-Level-7' into branch-Level-8

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Event.java

xuwen-ho committed:
Subject: committing hal.txt file to allow for merging

xuwen-ho committed:
Subject: Added classes Main.java and TaskList.java to increase OOP nature of code

Current State: Everything runs through UI.java, which contain the main()

What was Wrong: Code can be further simplifies through using more different classes

How Commit changes Current State: Lets create more classes so that code is more readable and adaptable to future changes

xuwen-ho committed:
Subject: Completed integration with Main.java and TaskList.java

Current State: Added classes Main.java and TaskList.java are not yet completed

What was Wrong: Code can be further simplified through using more different classes

How Commit changes Current State: Lets create more classes so that code is more readable and adaptable to future changes

xuwen-ho committed:
Subject: Added class Parser.java to increase OOP nature of code

Current State: Ui.java does most of the parsing of user input

What was Wrong: Code can be further simplifies through using more different classes

How Commit changes Current State: Lets create more classes so that code is more readable and adaptable to future changes

xuwen-ho committed:
Subject: Added class Parser.java to increase OOP nature of code

Current State: Ui.java does most of the parsing of user input

What was Wrong: Code can be further simplifies through using more different classes

How Commit changes Current State: Lets create more classes so that code is more readable and adaptable to future changes

xuwen-ho committed:
Subject: Cleaned up all classes by removing commented lines and unncessary methods

Current State: Many comments, making code unreadable

xuwen-ho committed:
Organised all java files into seperate packages based on their use case

Previously, all files were in the same folder

This made things a bit unorganised

Let's organise all java classes into pcakages so things can be clearer

xuwen-ho committed:
Merge remote-tracking branch 'origin/add-gradle-support'

xuwen-ho committed:
Added checkstyle to build.gradle

Previously, used default gradle settings

Problematic as I needed more functionality from gradle

Lets add more things to gradle so we can have greater functionality

xuwen-ho committed:
Add JUnit Tests

Currently, no unit tests present.

Menas that we are unable to ensure correctness of non-trivial methods.

Lets add JUnit tests so as to automate the testing process and ensure code correctness.

xuwen-ho committed:
Add JAR functionality within gradle

Previously gradle was not configured to allow jar usage

This meant that we were not able to easily execute our program as a jar file

Lets make it easier to run our program as a jar file

xuwen-ho committed:
Add javadocs to the Main class

Previously, no javadocs were present

This made the main class and its constituent methods difficult to understand their usage

Lets make classes and their methods clearer in terms of their usage

xuwen-ho committed:
Fix files according to coding convention

Previously, there were some violations to coding conventions.

These affected readibility, mainly using all Caps for constants, as well as having indentations for cases.

Lets imporve code readability by following conventions.

xuwen-ho committed:
Add find command

Previously, we were unable to find tasks based on a keyword

This made it difficult to see which tasks we had left based on a specific subject

Lets add the find command so that we can find relevant tasks much quicker

xuwen-ho committed:
Merge branch 'Level-9'

xuwen-ho committed:
Merge branch 'A-JavaDoc'

agreatdayy commented on src/main/java/duke/gui/Ui.java:
```diff
@@ -0,0 +1,97 @@
+import java.util.Scanner;
+
+
+public class Ui {
```
Should the variables in this class be privated or protected?

agreatdayy commented on src/main/java/duke/gui/Ui.java:
```diff
@@ -0,0 +1,97 @@
+                String taskString = taskList.addTask(userInput);
+
+                // If there is no error, display the messages that task has been added
+                if (!Objects.equals(taskString, "error")) {
```
Is there a more intuitive way of checking if `taskString` equals to `"error"`?

agreatdayy commented on src/main/java/duke/gui/Ui.java:
```diff
@@ -0,0 +1,97 @@
+        }
+
+        System.out.println(divider);
+        System.out.println("This mission is too important for me to allow you to jeopardize it. Goodbye.");
```
I like how your chatbot makes its grand exit! 😄

chinzhening replied:
I agree! Awesome feature 😮🌞

agreatdayy commented on src/main/java/duke/task/Storage.java:
```diff
@@ -0,0 +1,77 @@
+                boolean isDone = (Integer.parseInt(parts[1]) == 1);
+                String description = parts[2];
+
+                switch (taskType) {
+                case "T":
+                    prevTaskList.add(new Todo(isDone, description));
+                    break;
+                case "D":
+                    prevTaskList.add(new Deadline(IS_DONE_DEFAULT, description, parts[3]));
+                    break;
+                case "E":
+                    prevTaskList.add(new Event(IS_DONE_DEFAULT, description, parts[3], parts[4]));
+                    break;
+                default:
+                    System.out.println("Unknown task type");
+                }
+            }
+        } catch (Exception e) {
+            System.out.println(e.getMessage());
+        }
+
```
I like how neat this part is 👍 Good job!!

agreatdayy commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,114 @@
+
+
+public class TaskList {
+        ArrayList<Task> taskList;
+        boolean IS_DONE_DEFAULT = false;
+        Parser parser;
```
Should these variables be made privated or protected?

agreatdayy commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,33 @@
+
+    protected LocalDate deadlineFrom;
+    protected LocalDate deadlineTo;
+    String DATE_FORMAT = "MMM d yyyy";
+    String DIVIDER = " | ";
```
Should this be made `static final`? I noticed the same issue in several other places too. Press on! 👍

agreatdayy reviewed (commented):
Overall, your code is very easy to follow. You can work on your javadocs and your code layout(e.g. add extra line in between methods) to make it even better! All the best! 👍

chinzhening commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,29 @@
+public class Deadline extends Task {
+
+    protected LocalDate by;
+    String DATE_FORMAT = "MMM d yyyy";
+    String DIVIDER = " | ";
```
Should these variables be static final? What about the access modifier?

... I noticed the same issue in several other places too

chinzhening commented on src/main/java/duke/task/Storage.java:
```diff
@@ -0,0 +1,77 @@
+import java.util.Scanner;
+
+public class Storage {
+    boolean IS_DONE_DEFAULT = false;
+    File file;
+    private final TaskList taskList;
+    public Storage(String filePath, TaskList taskList) {
```
Perhaps you can add some breaks between these lines for better readability?

chinzhening commented on src/main/java/duke/gui/Main.java:
```diff
@@ -0,0 +1,56 @@
+ */
+public class Main {
+
+    private Storage storage;
+    private TaskList taskList;
+    private Ui ui;
+    private String filePath;
```
Should these be final?

chinzhening commented on src/main/java/duke/gui/Ui.java:
```diff
@@ -0,0 +1,97 @@
+                    System.out.printf("Now you have %d tasks in the list.", taskList.getNumberOfTasks());
+                }
+
+            } else if (userInputArray[0].equalsIgnoreCase("delete")) {
```
Is it an intended feature for the command to not be case-sensitive? What is the rationale for this?

chinzhening reviewed (commented):
LGTM, only a few small things to fix.

xuwen-ho committed:
Add some GUI features

Previously, these were not present

Only present on CLI, which is not user-friendly

Lets make our chatbot easier to use

xuwen-ho committed:
Add some GUI features integrated with FXML

Previously, these were not present

Only present on CLI, which is not user-friendly

Lets make our chatbot easier to use

xuwen-ho committed:
Implment GUI so that we can implment more commands

Previously, GUI was only able to echo commands

This made it difficult to use GUI

Lets make the UI more counter intuitive

xuwen-ho committed:
Add assertion in TaskList.java

Previously, error checking only ensures dates were in correct format

Meant that errors pertaining to tasks without descriptions were not caught

Lets catch errors where there is no task description

xuwen-ho committed:
Add code quality improvements to 2 classes, Ui.java and TaskList.java

Previously, best practices were not followed, leading to poor readability

xuwen-ho committed:
Merge pull request #2 from xuwen-ho/branch-A-Assertions

Add assertion in TaskList.java

xuwen-ho committed:
Merge branch 'master' into branch-A-CodeQuality

xuwen-ho committed:
Merge pull request #3 from xuwen-ho/branch-A-CodeQuality

Branch a code quality

xuwen-ho committed:
Merge branch 'branch-A-CodeQuality'

xuwen-ho committed:
Add help command to list all possible commands available

Previously, we do not know which commands we can run

Now we have the whole list of commands provided

Lets make it clear to the user what commands can be used

xuwen-ho committed:
Merge branch 'master' into branch-A-CodeQuality

xuwen-ho committed:
Merge branch 'branch-BCD-Extension'

xuwen-ho committed:
Merge branch 'master' of https://github.com/xuwen-ho/ip

xuwen-ho committed:
Add User Guide

Previously there was no user guide, meaning users do not know which commands to use

xuwen-ho committed:
Fix image not appearing in README

xuwen-ho committed:
Add JavaDocs for all classes

Preiovusly, there was no documentation for all classes

This made it difficult to understand the purpose of the methods

Lets make it clear what each class and method is about

xuwen-ho committed:
Merge branch 'branch-A-CodeQuality'

xuwen-ho committed:
Improve GUI by adding to FXML files

Previously, user interface was very basic

Now, it is more colourful, looking like an actual chatbot

Lets make our chatbot more user friendly!

xuwen-ho committed:
Fix several bugs, relating to the build.gradle file and Storage system

Initially, Storage system failed to store whether or not Deadlines and Events were marked

This led to a loss of knowledge about whether these Tasks were marked or not

Lets ensure we properly track each and every tasks' status

xuwen-ho committed:
Update README
